### DISCUSSION TITLE
Is there a correct way to pass context data to each query?

### DISCUSSION THREAD

**Discussion by Naggertooth** [prisma/prisma#20907] [Q&A] [answered]
### Question

I was looking for any right implementation of passing custom arguments to queries but there are only one documented way to do this with prisma (as i have found per current moment)
https://github.com/prisma/prisma-client-extensions/blob/4b7fe4549121a8d1eb73499230f84dd1b91733ba/row-level-security/script.ts#L21-L37

And i don't want to pass context data on each prisma initialization for each user, i'd be more happy if there would be a way to add custom arguments, like:
```typescript
export const extendedPrismaClient = prisma.$extends({
  query: {
    user: {
      async $allOperations({ model, operation, args, query /* or any other way to pass custom metadata */ }) {
        const conditionMissingContextUser = !(
          args as typeof args & { byUser: UserModel }
        ).byUser;
        if (conditionMissingContextUser) {
          throw new Error('Extended operation missing context user.');
        }

        const contextUser = (args as typeof args & { byUser: UserModel })
          .byUser;

        args = {
          ...args,
          where: {
            ...mergePermissionCheckFilterPart(
              (args as Extract<typeof args, { where: unknown }>).where,
              await sharedWhereFilterPartToMerge(contextUser)
            ),
          },
        } as Extract<typeof args, { where: unknown }>;

        return query(args);
      },
    },
  },
});
```

**Comment by Naggertooth** [comment]
Hi again!

I could be moving in the wrong direction.
Or if I've confused the category of the question or if I'm missing something important from the question or need to provide more context - please, let me know about that somehow.

**Reply by ludralph** [reply]
Hi @Naggertooth, I've shared your question internally with the engineering team. I'll get back to you with any updates

**Reply by Naggertooth** [reply]
Hi! Nice! Thank you very much!

**Reply by millsp** [reply]
Hey @Naggertooth, are you looking for a way to pass a specifc context per query as an extra argument (eg. `prisma.user.findFirst({ ..., context })` or something different?

**Reply by Naggertooth** [reply]
Hi @millsp, I guess I am looking for something different.

I am trying to pass an additional argument to each method in the prisma.$extends instance I created.
Unfortunately, I can't use the create option: 
```function forUser(userId: string) { return Prisma.defineExtension((prisma) => prisma.$extends(...))) }```,
because on every invocation where I am forced to filter data for a user - I will be forced to create a new prisma.$extends instance, and that doesn't seem like a good idea to me. Instead, I'd like to be able to pass an extra argument to each method of the prisma.$extends instance I've created.

It might also be worth noting that I'm using "nestjs-prisma", and if it's possible to do what I want - I won't have to figure out: how to do dynamic creation of extendedPrismaClient with passing an extra context in the argument. (But absolutely I won't necessarily use such a solution, there are several other alternatives at the moment, like: fetching context from a global temporary datastore and a couple others. )

But all of these solutions don't seem right to me. That's why I decided to turn here.

**Reply by millsp** [reply, marked as answer]
I see, I don't think I have a proper solution for you right now. Could you please create a feature/improvement request with the details?
Also, fwiw, extensions are very cheap, so you can actually use `$extends` at will, it does not create a new instance it reuses the existing one.

**Reply by Naggertooth** [reply]
Thanks for help! Absolutely, I'll create feature request with the details. And thanks for mentioning that extensions are cheap! This is game-changing thing for me.

**Reply by KhaledAlMana** [reply]
@Naggertooth 

Have been going around with no solution, especially when using a query instead of a model.
Til I figured this out.
[nestjs-prisma-auditable-ext.ts](https://gist.github.com/KhaledAlMana/f6a79fe6eb0a39afe2c872737b803a62)
Caution: Haven't yet used it in production.

**Comment by moisesrms** [comment]
This might be useful

https://medium.com/streamdal/creating-a-prisma-client-extension-27bb1aaabb0e

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
